lib/logstash/inputs/snmp/clientv3.rb

```diff
@@ -57,13 +45,29 @@ def initialize(protocol, address, port, retries, timeout, mib, security_name, au
       auth_pass = auth_pass.nil? ? nil : OctetString.new(auth_pass)
       priv_pass = priv_pass.nil? ? nil : OctetString.new(priv_pass)
 
-      # make sure the USM is initialized
-      USMFactory.instance
+      engine_id = OctetString.new(MPv3.createLocalEngineID)
+
+      security_protocols = SecurityProtocols.getInstance
```

Comment:
We're still using the singleton instances of `SecurityProtocols` and `SecurityModels` here, which is surprising to me. Part of the reason I previously recommended initializing our `org.snmp4j.Snmp` with an explicit `MessageDispatcher` was so that we could avoid implicitly referencing these singletons.

Reply:
`SecurityProtocols` is a singleton class. For `SecurityModels` I haven't found a need to create a new instance. The important part here which correctly associate the `usm` to support requests using the same `security_name` on different hosts using different credentials is:

```
dispatcher.addMessageProcessingModel(MPv3.new(usm))
```

Where the `usm` is associated with the `MPv3`.